File: packages/python/lizard/platform/addons.py
from __future__ import annotations
from dataclasses import dataclass
from typing import TYPE_CHECKING, Literal

if TYPE_CHECKING:
    from .client import PlatformClient

AddonType = Literal["postgres", "mysql", "mongodb", "redis", "s3"]


@dataclass
class Addon:
    id: str
    name: str
    type: str
    project_id: str
    status: str
    region: str | None = None
    version: str | None = None
    storage_mi: int | None = None
    memory_mi: int | None = None
    cpu_millis: int | None = None

    @classmethod
    def _from_dict(cls, d: dict) -> "Addon":
        return cls(
            id=d["id"],
            name=d["name"],
            type=d.get("type", ""),
            project_id=d.get("projectId", ""),
            status=d.get("status", "none"),
            region=d.get("region"),
            version=d.get("version"),
            storage_mi=d.get("storageMi"),
            memory_mi=d.get("memoryMi"),
            cpu_millis=d.get("cpuMillis"),
        )


class AddonsAPI:
    def __init__(self, client: "PlatformClient") -> None:
        self._client = client

    def list(self, *, project_id: str) -> list[Addon]:
        """List all addons in a project."""
        return [Addon._from_dict(a) for a in self._client.get(f"/api/projects/{project_id}/addons")]

    def get(self, project_id: str, addon_id: str) -> Addon:
        """Get an addon by project + addon ID."""
        return Addon._from_dict(self._client.get(f"/api/projects/{project_id}/addons/{addon_id}"))

    def create(
        self,
        *,
        project_id: str,
        name: str,
        type: AddonType,
        version: str | None = None,
        storage_mi: int | None = None,
        memory_mi: int | None = None,
        cpu_millis: int | None = None,
        region: str | None = None,
    ) -> Addon:
        """Create an addon (database, cache, or object storage)."""
        body = {k: v for k, v in {
            "name": name,
            "type": type,
            "version": version,
            "storageMi": storage_mi,
            "memoryMi": memory_mi,
            "cpuMillis": cpu_millis,
            "region": region,
        }.items() if v is not None}
        return Addon._from_dict(self._client.post(f"/api/projects/{project_id}/addons", body))

    def delete(self, project_id: str, addon_id: str) -> None:
        """Delete an addon."""
        self._client.delete(f"/api/projects/{project_id}/addons/{addon_id}")

    def resize(
        self,
        project_id: str,
        addon_id: str,
        *,
        storage_mi: int | None = None,
        memory_mi: int | None = None,
        cpu_millis: int | None = None,
    ) -> Addon:
        """Resize an addon's resources."""
        body = {k: v for k, v in {
            "storageMi": storage_mi,
            "memoryMi": memory_mi,
            "cpuMillis": cpu_millis,
        }.items() if v is not None}
        return Addon._from_dict(self._client.patch(f"/api/projects/{project_id}/addons/{addon_id}", body))

    def redeploy(self, project_id: str, addon_id: str) -> None:
        """Trigger a redeploy of an addon."""
        self._client.post(f"/api/projects/{project_id}/addons/{addon_id}/redeploy")

File: packages/js/src/platform/addons.ts
import type { PlatformClient } from './client'

export type AddonType = 'postgres' | 'mysql' | 'mongodb' | 'redis' | 's3'

export interface Addon {
  id: string
  name: string
  type: AddonType
  projectId: string
  status: 'none' | 'running' | 'crashed' | 'stopped'
  deployStatus: string
  version?: string
  /** Exposed connection variables (e.g. DATABASE_URL, REDIS_URL). */
  env?: Record<string, string>
}

export interface CreateAddonOpts {
  projectId: string
  type: AddonType
  name?: string
  version?: string
  /** vCPU count */
  vcpu?: number
  /** Memory in MB */
  memoryMb?: number
  /** Storage in GB */
  storageGb?: number
}

export class AddonsAPI {
  constructor(private readonly client: PlatformClient) {}

  /** List all addons in a project. */
  list(opts: { projectId: string }): Promise<Addon[]> {
    return this.client.get(`/api/projects/${opts.projectId}/addons`)
  }

  /** Get an addon by ID. */
  get(projectId: string, addonId: string): Promise<Addon> {
    return this.client.get(`/api/projects/${projectId}/addons/${addonId}`)
  }

  /**
   * Create a new managed addon (database, cache, or object store).
   *
   * @example
   * ```ts
   * const pg = await lizard.addons.create({ projectId, type: 'postgres' })
   * // Inject into a service:
   * await lizard.secrets.set({ serviceId: svcId, key: 'DATABASE_URL', value: `${{${pg.name}.DATABASE_URL}}` })
   * ```
   */
  create(opts: CreateAddonOpts): Promise<Addon> {
    return this.client.post(`/api/projects/${opts.projectId}/addons`, {
      type: opts.type,
      name: opts.name,
      version: opts.version,
      vcpu: opts.vcpu,
      memoryMb: opts.memoryMb,
      storageGb: opts.storageGb,
    })
  }

  /** Delete an addon. */
  delete(projectId: string, addonId: string): Promise<void> {
    return this.client.delete(`/api/projects/${projectId}/addons/${addonId}`)
  }

  /** Resize an addon (CPU / memory / storage). */
  resize(projectId: string, addonId: string, opts: { vcpu?: number; memoryMb?: number; storageGb?: number }): Promise<Addon> {
    return this.client.post(`/api/projects/${projectId}/addons/${addonId}/resize`, opts)
  }

  /** Restart an addon VM. */
  redeploy(projectId: string, addonId: string): Promise<void> {
    return this.client.post(`/api/projects/${projectId}/addons/${addonId}/redeploy`, {})
  }
}
